dev/buildtool/spinnaker_commands.py

```diff
@@ -0,0 +1,216 @@
+
+    logging.info('Creating and pushing branch "%s" to "%s"',
+                 branch, repository.origin)
+    self.__git.check_run(git_dir, 'checkout -b ' + branch)
+    self.__git.push_branch_to_origin(git_dir, branch)
+
+
+class PublishSpinnakerFactory(CommandFactory):
+  """"Implements the publish_spinnaker command."""
+  def __init__(self):
+    super(PublishSpinnakerFactory, self).__init__(
+        'publish_spinnaker', PublishSpinnakerCommand,
+        'Publish a spinnaker release')
+
+  def init_argparser(self, parser, defaults):
+    super(PublishSpinnakerFactory, self).init_argparser(parser, defaults)
+    HalRunner.add_parser_args(parser, defaults)
+    GitRunner.add_parser_args(parser, defaults)
+    GitRunner.add_publishing_parser_args(parser, defaults)
+
+    self.add_argument(
+        parser, 'spinnaker_version', defaults, None,
+        help='The spinnaker version to publish.')
+    self.add_argument(
+        parser, 'spinnaker_release_alias', defaults, None, required=True,
+        help='The spinnaker version alias to publish as.')
+    self.add_argument(
+        parser, 'halyard_bom_bucket', defaults, 'halconfig',
+        help='The bucket manaing halyard BOMs and config profiles.')
+    self.add_argument(
+        parser, 'bom_version', defaults, None,
+        help='The existing bom version usef for this release.')
+    self.add_argument(
+        parser, 'min_halyard_version', defaults, None,
+        help='The minimum halyard version required.')
+
+
+class PublishSpinnakerCommand(CommandProcessor):
+  """"Implements the publish_spinnaker command."""
+  # pylint: disable=too-few-public-methods
+
+  def __init__(self, factory, options, **kwargs):
+    super(PublishSpinnakerCommand, self).__init__(factory, options, **kwargs)
+    check_options_set(options, [
+        'spinnaker_version',
+        'bom_version',
+        'github_owner',
+        'min_halyard_version'
+    ])
+
+    options_copy = copy.copy(options)
+    self.__scm = BomSourceCodeManager(options_copy, self.get_input_dir())
+    self.__hal = HalRunner(options)
+    self.__git = GitRunner(options)
+    self.__hal.check_property(
+        'spinnaker.config.input.bucket', options.halyard_bom_bucket)
+
+  def push_branches_and_tags(self, bom):
+    """Update the release branches and tags in each of the BOM repositires."""
+    major, minor, _ = self.options.spinnaker_version.split('.')
+    branch = 'release-{major}.{minor}.x'.format(major=major, minor=minor)
+    logging.info('Tagging each of the BOM service repos')
+
+    # Run in two passes so we dont push anything if we hit a problem
+    # in the tagging pass. Since we are spread against multiple repositiories,
+    # we cannot do this atomically. The two passes gives us more protection
+    # from a partial push due to errors in a repo.
+    for which in ['tag', 'push']:
+      for name, spec in bom['services'].items():
+        if name in ['monitoring-third-party', 'defaultArtifact']:
```

Comment:
Where does `defaultArtifact` come from here? Just curious.

Reply:
I'm not sure. I saw these in a bom retrieved from halyard. It just occurred to me that maybe I accidently put it in there. I'll have to investigate.